File: leetcode-go/Q051-Q100/Q084_Largest_Rectangle_in_Histogram_B/Q084_Largest_Rectangle_in_Histogram_B.go
package Q084_Largest_Rectangle_in_Histogram_B

func largestRectangleArea(heights []int) int {
	stack := []int{}
	ans := 0

	i := 0
	for i < len(heights) {
		if len(stack) == 0 || heights[i] >= heights[stack[len(stack)-1]] {
			stack = append(stack, i)
			i++
			continue
		}

		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		w := i
		if len(stack) != 0 {
			w = i - stack[len(stack)-1] - 1
		}
		ans = max(ans, w*heights[top])
	}

	for len(stack) != 0 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		w := i
		if len(stack) != 0 {
			w = i - stack[len(stack)-1] - 1
		}
		ans = max(ans, w*heights[top])
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: src/Y2018/range_query/Q084_Largest_Rectangle_in_Histogram_B.java
package Y2018.range_query;

import java.util.Stack;

public class Q084_Largest_Rectangle_in_Histogram_B {

    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);
        int result = 0;
        for (int i = 0; i < heights.length; ++i) {
            while (stack.peek() != -1 && heights[stack.peek()] >= heights[i]) {
                result = Math.max(result, heights[stack.pop()] * (i - stack.peek() - 1));
            }
            stack.push(i);
        }
        while (stack.peek() != -1) {
            result = Math.max(result, heights[stack.pop()] * (heights.length - stack.peek() - 1));
        }
        return result;
    }

}
